src/mlpack/core/optimizers/katyusha/katyusha_impl.hpp

```diff
@@ -0,0 +1,188 @@
+  arma::mat z = arma::zeros<arma::mat>(iterate.n_rows, iterate.n_cols);
+  arma::mat w = arma::zeros<arma::mat>(iterate.n_rows, iterate.n_cols);
+
+  // Find the number of batches.
+  size_t numBatches = numFunctions / batchSize;
+  if (numFunctions % batchSize != 0)
+    ++numBatches; // Capture last few.
+
+  const size_t actualMaxIterations = (maxIterations == 0) ?
+      std::numeric_limits<size_t>::max() : maxIterations;
+  for (size_t i = 0; i < actualMaxIterations; /* incrementing done manually */)
+  {
+    // Is this iteration the start of a sequence?
+    if ((currentFunction % numFunctions) == 0)
+    {
+      // Output current objective function.
+      Log::Info << "Katyusha: iteration " << i << ", objective "
+          << overallObjective << "." << std::endl;
+
+      if (std::isnan(overallObjective) || std::isinf(overallObjective))
+      {
+        Log::Warn << "Katyusha: converged to " << overallObjective
+            << "; terminating  with failure.  Try a smaller step size?"
+            << std::endl;
+        return overallObjective;
+      }
+
+      if (std::abs(lastObjective - overallObjective) < tolerance)
+      {
+        Log::Info << "Katyusha: minimized within tolerance " << tolerance
+            << "; terminating optimization." << std::endl;
+        return overallObjective;
+      }
+
+      // Reset the counter variables.
+      lastObjective = overallObjective;
+      overallObjective = 0;
+      currentFunction = 0;
+
+      if (shuffle) // Determine order of visitation.
+        function.Shuffle();
+    }
+
+    iterate0 = iterate;
+    y = iterate;
+    z = iterate;
+    w.zeros();
+
+    // Compute the full gradient.
+    size_t effectiveBatchSize = std::min(batchSize, numFunctions);
+    function.Gradient(iterate, 0, fullGradient, effectiveBatchSize);
+    for (size_t f = effectiveBatchSize; f < numFunctions;
+        /* incrementing done manually */)
+    {
+      // Find the effective batch size (the last batch may be smaller).
+      effectiveBatchSize = std::min(batchSize, numFunctions - f);
+
+      function.Gradient(iterate, f, gradient, effectiveBatchSize);
+      fullGradient += gradient;
+
+      f += effectiveBatchSize;
+    }
+    fullGradient /= (double) numFunctions;
+
+    // To keep track of where we are and how things are going.
+    double bb = 0;
+    size_t iteration = 0;
+
+    for (size_t f = 0; f < numFunctions; /* incrementing done manually */)
+    {
+      // Find the effective batch size (the last batch may be smaller).
+      effectiveBatchSize = std::min(batchSize, numFunctions - f);
+
+      iterate = tau * z + 0.5 * iterate0 + y * tau1;
+
+      // Calculate variance reduced gradient.
+      function.Gradient(iterate, f, gradient, effectiveBatchSize);
+      function.Gradient(iterate0, f, gradient0, effectiveBatchSize);
+
+      const arma::mat z0 = z;
+      z -= stepSize * (fullGradient + (gradient - gradient0) /
+        (double) batchSize);
```

Comment:
Hmmm, I am not sure on this one but it might be possible to avoid an extra allocation and copy here (depending on how Armadillo handles the `z -= ...` line:

```
arma::mat zNew = z - stepSize * (fullGradient + (gradient - gradient0) / (double) batchSize);
y = iterate + (zNew - z) * tau;
z = std::move(zNew);
```

But I am not sure if it is actually faster (or if this is a bottleneck in the algorithm), so I dunno if it is worth trying.